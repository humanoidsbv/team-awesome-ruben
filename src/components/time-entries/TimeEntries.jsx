import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

//* todo: add icon to select
// import IconArrowDown from '../../assets/icons/icon-arrow-down.svg';
import styles from './TimeEntries.module.css';
import TimeEntry from '../time-entry/';
import TimeEntryAdd from '../time-entry-add/';
import TimeEntryHeader from '../time-entry-header/';

const TimeEntries = ({
  addTimeEntry,
  clients,
  deleteTimeEntry,
  fetchClients,
  fetchTimeEntries,
  filterTimeEntriesByClient,
  timeEntries
}) => {
  useEffect(function getEntryData() {
    fetchClients();
    fetchTimeEntries();
  }, []);

  //* Note: html select elements return strings while we're working with an id (number)
  const handleChange = event =>
    filterTimeEntriesByClient(
      !event.target.value ? null : Number(event.target.value)
    );

  return (
    <React.Fragment>
      <TimeEntryAdd addFormData={addTimeEntry} clients={clients} />
      <div className={styles.header}>
        <span className={styles.heading}>All time entries</span>
        <select
          className={styles.sortMembers}
          onChange={handleChange}
          type="button"
        >
          <option value="">No filter</option>
          {clients.map(({ companyName, id }) => (
            <option value={id} key={id}>
              {companyName}
            </option>
          ))}
        </select>
      </div>
      {timeEntries.map(
        ({ startTimestamp, stopTimestamp, id, client }, index) => {
          const previousItem = timeEntries[index - 1];
          const currentDate = new Date(startTimestamp).toDateString();
          const previousDate = previousItem
            ? new Date(previousItem.startTimestamp).toDateString()
            : '';

          return (
            <React.Fragment key={id}>
              {currentDate !== previousDate && (
                <TimeEntryHeader
                  startTimestamp={startTimestamp}
                  stopTimestamp={stopTimestamp}
                />
              )}
              <TimeEntry
                client={client.companyName}
                deleteEntry={timeEntryId => deleteTimeEntry(timeEntryId)}
                startTimestamp={startTimestamp}
                stopTimestamp={stopTimestamp}
                timeEntryId={id}
              />
            </React.Fragment>
          );
        }
      )}
    </React.Fragment>
  );
};

TimeEntries.propTypes = {
  addTimeEntry: PropTypes.func.isRequired,
  clients: PropTypes.arrayOf(
    PropTypes.shape({
      companyName: PropTypes.string
    })
  ).isRequired,
  deleteTimeEntry: PropTypes.func.isRequired,
  fetchClients: PropTypes.func.isRequired,
  fetchTimeEntries: PropTypes.func.isRequired,
  filterTimeEntriesByClient: PropTypes.func.isRequired,
  timeEntries: PropTypes.arrayOf(
    PropTypes.shape({
      startTimestamp: PropTypes.string
    })
  ).isRequired
};

export default TimeEntries;
