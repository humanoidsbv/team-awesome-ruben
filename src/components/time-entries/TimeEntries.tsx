import React, { useEffect } from 'react';

import styles from './TimeEntries.module.css';
import TimeEntry from '../time-entry';
import TimeEntryAdd from '../time-entry-add/';
import TimeEntryHeader from '../time-entry-header/';

import { ClientInterface } from '../../redux/clients/types';
import { TimeEntryInterface } from '../../redux/time-entries/types';

export interface TimeEntriesProps {
  addTimeEntry: () => void;
  clients: ClientInterface[];
  deleteTimeEntry: (timeEntryId: number) => void;
  fetchClients: () => void;
  fetchTimeEntries: () => void;
  filterTimeEntriesByClient: (eventTargetValue: number) => void;
  timeEntries: TimeEntryInterface[];
  isLoading: boolean;
}

const TimeEntries = ({
  addTimeEntry,
  clients,
  deleteTimeEntry,
  fetchClients,
  fetchTimeEntries,
  filterTimeEntriesByClient,
  isLoading,
  timeEntries
}: TimeEntriesProps): React.ReactElement => {
  useEffect(() => {
    fetchClients();
    fetchTimeEntries();
  }, []);

  //* Note: a select element returns a string. We're working with a number
  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>): void =>
    filterTimeEntriesByClient(
      !event.target.value ? null : Number(event.target.value)
    );

  return (
    <React.Fragment>
      <TimeEntryAdd addFormData={addTimeEntry} clients={clients} />
      <div className={styles.header}>
        <span className={styles.heading}>All time entries</span>
        <select className={styles.sortMembers} onChange={handleChange}>
          <option value="">No filter</option>
          {clients.map(({ companyName, id }) => (
            <option value={id} key={id}>
              {companyName}
            </option>
          ))}
        </select>
      </div>
      {!isLoading &&
        timeEntries.map(
          ({ startTimestamp, stopTimestamp, id, client }, index: number) => {
            const previousItem = timeEntries[index - 1];
            const currentDate = new Date(startTimestamp).toDateString();
            const previousDate = previousItem
              ? new Date(previousItem.startTimestamp).toDateString()
              : '';

            return (
              <React.Fragment key={id}>
                {currentDate !== previousDate && (
                  <TimeEntryHeader
                    startTimestamp={startTimestamp}
                    stopTimestamp={stopTimestamp}
                  />
                )}
                <TimeEntry
                  client={client}
                  deleteEntry={deleteTimeEntry}
                  startTimestamp={startTimestamp}
                  stopTimestamp={stopTimestamp}
                  id={id}
                />
              </React.Fragment>
            );
          }
        )}
    </React.Fragment>
  );
};

export default TimeEntries;
