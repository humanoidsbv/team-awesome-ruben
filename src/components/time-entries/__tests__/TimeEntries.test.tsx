import * as React from 'react';
import { shallow } from 'enzyme';

import TimeEntries from '../TimeEntries';

const mockedClientsData = {
  branch: '',
  branchAddress: '',
  companyName: '',
  id: 0,
  locality: '',
  postalCode: ''
};

const mockedTimeEntriesData = {
  client: mockedClientsData,
  id: 0,
  startTimestamp: '',
  stopTimestamp: ''
};

it('matches output with  snapshot', () => {
  const timeEntries = shallow(
    <TimeEntries
      addTimeEntry={() => {}}
      deleteTimeEntry={() => {}}
      fetchTimeEntries={() => {}}
      fetchClients={() => {}}
      timeEntries={[mockedTimeEntriesData]}
      clients={[mockedClientsData]}
      filterTimeEntriesByClient={() => {}}
    />
  );
  expect(timeEntries).toMatchSnapshot();
});
