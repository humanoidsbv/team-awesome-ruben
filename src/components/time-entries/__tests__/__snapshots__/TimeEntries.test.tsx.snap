// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches output with  snapshot 1`] = `
<Fragment>
  <TimeEntryAdd
    addFormData={[Function]}
    clients={
      Array [
        Object {
          "branch": "",
          "branchAddress": "",
          "companyName": "",
          "id": 0,
          "locality": "",
          "postalCode": "",
        },
      ]
    }
  />
  <div
    className="header"
  >
    <span
      className="heading"
    >
      All time entries
    </span>
    <select
      className="sortMembers"
      onChange={[Function]}
    >
      <option
        value=""
      >
        No filter
      </option>
      <option
        key="0"
        value={0}
      />
    </select>
  </div>
  <TimeEntryHeader
    startTimestamp=""
    stopTimestamp=""
  />
  <TimeEntry
    client={
      Object {
        "branch": "",
        "branchAddress": "",
        "companyName": "",
        "id": 0,
        "locality": "",
        "postalCode": "",
      }
    }
    deleteEntry={[Function]}
    id={0}
    startTimestamp=""
    stopTimestamp=""
  />
</Fragment>
`;
